shared/typeflow/codeql/typeflow/internal/TypeFlowImpl.qll

```diff
@@ -0,0 +1,456 @@
+      )
+    }
+
+    int lastRank(Node n) { result = max(int r | edgeRank(r, _, n)) }
+  }
+
+  private signature module TypePropagation {
+    class Typ;
+
+    predicate candType(TypeFlowNode n, Typ t);
+
+    bindingset[t]
+    predicate supportsType(TypeFlowNode n, Typ t);
+  }
+
+  /** Implements recursion through `forall` by way of edge ranking. */
+  private module ForAll<NodeSig Node, RankedEdge<Node> E, TypePropagation T> {
+    /**
+     * Holds if `t` is a bound that holds on one of the incoming edges to `n` and
+     * thus is a candidate bound for `n`.
+     */
+    pragma[nomagic]
+    private predicate candJoinType(Node n, T::Typ t) {
+      exists(TypeFlowNode mid |
+        T::candType(mid, t) and
+        E::edgeRank(_, mid, n)
+      )
+    }
+
+    /**
+     * Holds if `t` is a candidate bound for `n` that is also valid for data coming
+     * through the edges into `n` ranked from `1` to `r`.
+     */
+    private predicate flowJoin(int r, Node n, T::Typ t) {
+      (
+        r = 1 and candJoinType(n, t)
+        or
+        flowJoin(r - 1, n, t) and E::edgeRank(r, _, n)
+      ) and
+      forall(TypeFlowNode mid | E::edgeRank(r, mid, n) | T::supportsType(mid, t))
+    }
+
+    /**
+     * Holds if `t` is a candidate bound for `n` that is also valid for data
+     * coming through all the incoming edges, and therefore is a valid bound for
+     * `n`.
+     */
+    predicate flowJoin(Node n, T::Typ t) { flowJoin(E::lastRank(n), n, t) }
+  }
+
+  private module JoinStep implements Edge {
+    class Node = TypeFlowNode;
+
+    predicate edge = joinStep/2;
+  }
+
+  private module SccJoinStep implements Edge {
+    class Node = TypeFlowScc;
+
+    predicate edge = sccJoinStep/2;
+  }
+
+  private module RankedJoinStep = RankEdge<JoinStep>;
+
+  private module RankedSccJoinStep = RankEdge<SccJoinStep>;
+
+  private module ExactTypePropagation implements TypePropagation {
+    class Typ = Type;
+
+    predicate candType = exactType/2;
+
+    predicate supportsType = exactType/2;
+  }
+
+  /**
+   * Holds if the runtime type of `n` is exactly `t` and if this bound is a
+   * non-trivial lower bound, that is, `t` has a subtype.
+   */
+  private predicate exactType(TypeFlowNode n, Type t) {
+    exactTypeBase(n, t)
+    or
+    exists(TypeFlowNode mid | exactType(mid, t) and step(mid, n))
+    or
+    // The following is an optimized version of
+    // `forex(TypeFlowNode mid | joinStep(mid, n) | exactType(mid, t))`
+    ForAll<TypeFlowNode, RankedJoinStep, ExactTypePropagation>::flowJoin(n, t)
+    or
+    exists(TypeFlowScc scc |
+      sccRepr(n, scc) and
+      // Optimized version of
+      // `forex(TypeFlowNode mid | sccJoinStep(mid, scc) | exactType(mid, t))`
+      ForAll<TypeFlowScc, RankedSccJoinStep, ExactTypePropagation>::flowJoin(scc, t)
+    )
+  }
+
+  /**
+   * Gets the source declaration of a direct supertype of this type, excluding itself.
+   */
+  private Type getASourceSupertype(Type t) {
```

Comment:
Likely needs `pragma[nomagic]`.

Reply:
Fixed in a2c29fe094190664f6899468c16d846a300d9654